From 77074ebcbfae315d9c71553713fdac0aaac8ebc3 Mon Sep 17 00:00:00 2001
From: Elena Popa <elena.popa@nxp.com>
Date: Tue, 24 Sep 2024 13:05:48 +0300
Subject: [PATCH] imx8m: reset ethernet phy in spl

Reset the Ethernet PHY for the i.MX 8M Family. To bring it up the operational
state in which Ethernet MAC can communicate with the PHY, this must be reset in
SPL, before starting the kernel, when Falcon Mode is enabled.
 
Signed-off-by: Elena Popa <elena.popa@nxp.com>
---
 common/spl/spl.c | 29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

diff --git a/common/spl/spl.c b/common/spl/spl.c
index b65c439e7aa..059203d1ecf 100644
--- a/common/spl/spl.c
+++ b/common/spl/spl.c
@@ -4,6 +4,9 @@
  * Texas Instruments, <www.ti.com>
  *
  * Aneesh V <aneesh@ti.com>
+ *
+ * Copyright 2024 NXP
+ *
  */
 
 #include <common.h>
@@ -43,10 +46,20 @@
 #include <bootcount.h>
 #include <wdt.h>
 #include <video.h>
+#include <asm/mach-imx/gpio.h>
+#include <linux/delay.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 DECLARE_BINMAN_MAGIC_SYM;
 
+#ifdef CONFIG_ARCH_IMX8M
+#define RESET_ETH0_GPIO IMX_GPIO_NR(4,22)
+#ifdef CONFIG_IMX8MP
+#define RESET_ETH1_GPIO IMX_GPIO_NR(4,2)
+#endif
+#define USDHC_GPIO_PAD_CTRL (PAD_CTL_HYS | PAD_CTL_DSE1)
+#endif
+
 u32 *boot_params_ptr = NULL;
 
 #if CONFIG_IS_ENABLED(BINMAN_UBOOT_SYMBOLS)
@@ -711,6 +724,22 @@ void board_init_r(gd_t *dummy1, ulong dummy2)
 		}
 	}
 
+#ifdef CONFIG_ARCH_IMX8M
+	/*reset eth*/
+	gpio_request(RESET_ETH0_GPIO, "reset_eth0_gpio");
+	gpio_direction_output(RESET_ETH0_GPIO, 0);
+	mdelay(1);
+	gpio_direction_output(RESET_ETH0_GPIO, 1);
+	mdelay(1);
+#ifdef CONFIG_IMX8MP
+	gpio_request(RESET_ETH1_GPIO, "reset_eth1_gpio");
+	gpio_direction_output(RESET_ETH1_GPIO, 0);
+	mdelay(1);
+	gpio_direction_output(RESET_ETH1_GPIO, 1);
+	mdelay(1);
+#endif
+#endif
+
 	if (CONFIG_IS_ENABLED(BOARD_INIT))
 		spl_board_init();
 
-- 
2.25.1

